x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/AbstractTransportSearchableSnapshotsAction.java

```diff
@@ -0,0 +1,118 @@
+    }
+
+    AbstractTransportSearchableSnapshotsAction(String actionName, ClusterService clusterService, TransportService transportService,
+                                               ActionFilters actionFilters, IndexNameExpressionResolver resolver,
+                                               Writeable.Reader<Request> request, String executor, IndicesService indicesService,
+                                               boolean canTripCircuitBreaker) {
+        super(actionName, clusterService, transportService, actionFilters, resolver, request, executor, canTripCircuitBreaker);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected ClusterBlockException checkGlobalBlock(ClusterState state, Request request) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+    }
+
+    @Override
+    protected ClusterBlockException checkRequestBlock(ClusterState state, Request request, String[] indices) {
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, indices);
+    }
+
+    @Override
+    protected ShardsIterator shards(ClusterState state, Request request, String[] concreteIndices) {
+        final List<String> searchableSnapshotIndices = new ArrayList<>();
+        for (String concreteIndex : concreteIndices) {
+            IndexMetaData indexMetaData = state.metaData().index(concreteIndex);
+            if (indexMetaData != null) {
+                Settings indexSettings = indexMetaData.getSettings();
+                if (INDEX_STORE_TYPE_SETTING.get(indexSettings).equals(SNAPSHOT_DIRECTORY_FACTORY_KEY)) {
+                    if (SNAPSHOT_CACHE_ENABLED_SETTING.get(indexSettings)) {
+                        searchableSnapshotIndices.add(concreteIndex);
+                    }
+                }
+            }
+        }
+        if (searchableSnapshotIndices.isEmpty()) {
+            throw new ResourceNotFoundException("No searchable snapshots indices found");
+        }
+        return state.routingTable().allShards(searchableSnapshotIndices.toArray(new String[0]));
+    }
+
+    @Override
+    protected ShardOperationResult shardOperation(Request request, ShardRouting shardRouting) throws IOException {
+        final IndexShard indexShard = indicesService.indexServiceSafe(shardRouting.index()).getShard(shardRouting.id());
+        final CacheDirectory cacheDirectory = unwrap(indexShard.store().directory());
+        assert cacheDirectory != null;
+        assert cacheDirectory.getShardId().equals(shardRouting.shardId());
+        return executeShardOperation(request, shardRouting, cacheDirectory);
+    }
+
+    protected abstract ShardOperationResult executeShardOperation(Request request, ShardRouting shardRouting,
+                                                                  CacheDirectory cacheDirectory) throws IOException;
+
+    @Nullable
+    private static CacheDirectory unwrap(Directory dir) {
```

Comment:
Maybe `unwrapCacheDirectory` to be a bit more specific about how far we are unwrapping?

(I realise this code existed beforehand, but I only just thought of this)

Reply:
Sure, I pushed 544f0565c315c317585db9961cb654f5e81efcac